Improve sort performance with DataViewModel sorting

turdusmerula opened the pull request:
I was suffering terrible performance issue when trying to sort a list with only a few thousand of elements and eventually figured out what was going on.

In wxDataViewCtrlInternal::BuildBranch when building a sorted list of nodes we end up by sorting the child node list **everytime a new is added**, resulting in a tremendous and growing exponentially waste of time when increasing the number of items.

What I made here to fix this issue is to ensure that we wait to insert the last child before attempting to make a sort, leaving with only one sort per node.

Commit:
Improve sort performance

Commit:
Corrected whitespace typo

Commit:
Corrected whitespace typo

Commit:
Corrected whitespace typo

Commit:
Corrected whitespace typo

vadz commented:
Thanks, the idea is definitely right, but I'm strongly allergic to the use of [bool params](https://www.wxwidgets.org/develop/coding-guidelines/#no_bool_params) and would really like to get rid of them here as they make the code less readable than it ought to be.

In fact, I'd rather avoid adding any parameters at all and just add some new
```cpp
void ResortIfNeeded()
{
    if (m_internal->ShouldBeSorted())
    {
        gs_internal = m_internal;
        m_children.Sort( &wxGtkTreeModelChildCmp );
    }
}
```
and call it after the loop as this seems simpler and more clear. Could you please try doing it like this? TIA!

turdusmerula commented:
That was my first idea but it's not quite easy, to my mind the sort is not made at the right place from the beginning. I come from wxPython and have not found a way to insert a bunch of items and trigger the sort at the end to avoid the performance issue.
I'll try to find a cleaner way to achieve this !

Commit:
Refarctored correction, removed boolean

turdusmerula commented:
@vadz I refactored the code, hope it fits better the requirement now

vadz committed:
Fix indentation

Don't use hard TABs.

vadz committed:
Initialize new m_allow_sort member

Its initial value is otherwise undefined.

vadz committed:
Rename the new sort-related methods

Don't use SetAllowSort() in one class and {Allow,Forbid}Sort() in
another one, both could be fine on their own but not together.

Also simplify code by forbidding sorting only if necessary (and add a
comment explaining why do we do it), which removes the need for
re-allowing it from two different places.

vadz committed:
Rename AllowSort() to FreezeSort()

This better reflects the temporary nature of forbidding sort.

Also add a comment describing this better.

vadz commented:
Thanks, this looks much better, but I've pushed a few corrections improving things further -- could you please test these changes with your application and let me know if you see any problems with them? TIA!

turdusmerula commented:
Tested ok !

The fix proposed solves a performance issue but it does not solve the background issue that the DataViewModel is not easily scalable, and it's even worse when used from wxPython. When used from wxPython the Sort from stl turns to be very slow due to the fact that it goes back to Python on each item compare, as items are stored as variants it divides performance by a non negligible factor.
When I have time I will try to propose some improvements:
 - implement a CachedDataviewModel that would use an indexed map to store childs allowing to maintain in parallel a sorted structure for each column. The items would then be maintained sorted at insertion / update / remove time allowing sort to be instantaneous from user point of view when clicking on column header. There is a cost at insertion time (that is already there by the way) but it's greater in terms of user experience. 
 - add helpers to enable lazy loading capabilities more easily (we already can but it looks hacky in some way) that when used would transfer the sorting responsibility to the data provider.

vadz commented:
Thanks for testing!

FWIW I think that going from C++ to Python and back for each comparison is definitely not going to be ideal. IMO the only solution is to either do sorting entirely in C++ (which, of course, would only work for simple sort criteria) or sort items in Python before inserting them.

vadz commented:
Merged, after squashing some commits, as fb136cf1541e327ef5be61039e6b06c01ee96be3, thanks again!
